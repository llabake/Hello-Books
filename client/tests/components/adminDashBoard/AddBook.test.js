import React from 'react';

import {
  AddBook,
  mapDispatchToProps,
  mapStateToProps
} from '../../../components/adminDashBoard/AddBook'
import userData from '../../mocks/userData';


jest.useFakeTimers();

function setup() {
  const props = {
    dispatch: jest.fn(),
    location: {
      state: {
        from: {}
      }
    },
    isAuthenticated: false,
    saveBook: jest.fn(),
    checkIsbnExist: jest.fn().mockImplementation(() => Promise.resolve(1)),
    handleChange: jest.fn(),
    handleSubmit: jest.fn(),
    handleBlur: jest.fn(),
    handleFileChange: jest.fn(),
    errors: {}
  };

  const state = {
    title: '',
    author: '',
    publishedYear: '',
    isbn: '',
    quantity: '',
    description: '',
    aboutAuthor: '',
    image: '',
    errors: {},
    isbnExist: '',
    saving: false,
    isValid: false,
    userExist: {},
    redirect: false,
    uploadedImage: null
  };
  const wrapper = shallow(<AddBook {...props} {...state} />);

  const event = {
    preventDefault: jest.fn(),
    target: {
      name: 'title',
      value: 'title'
    }
  };

  return {
    wrapper,
    props,
    state,
    event
  };
}

describe('Add book component', () => {
  test('should render appropriately', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  })

  test('should change state for title', () => {
    const { wrapper, props, event, state } = setup();
    wrapper.find('.signup').simulate('change', event);
    wrapper.instance().handleChange(event)
    expect(wrapper.instance().state.title).toBe('title');
  });

  test('should call book exist to check if book exists before adding book', () => {
    const { wrapper, props, state } = setup();
    const handleBlurSpy = jest.spyOn(AddBook.prototype, 'handleBlur')
    const event = {
      preventDefault: jest.fn(),
      target: {
        name: 'isbn',
        value: '9871645327878'
      }
    }
    wrapper.setState({
      isbn: '9871645327878',
      // isbnExist: {
      //   isbn: 'isbn error'
      // }
    });
    wrapper.instance().handleBlur(event)
    expect(wrapper.instance().state.isbn).toBe('9871645327878');
    expect(props.checkIsbnExist).toHaveBeenCalled()
  });

  test('should handle form submission when the form inputs are valid', () => {
    const { state } = setup();
    const event = {
      preventDefault: jest.fn(),
      target: {}
    }

    const props = {
      dispatch: jest.fn(),
      user: userData.admin,
      errors: {},
      loading: false,
      updateBook: jest.fn(),
      checkIsbnExist: jest.fn().mockImplementation(() => Promise.resolve(1)),
      handleChange: jest.fn(),
      handleSubmit: jest.fn(),
      handleBlur: jest.fn(),
      book: {}
    };
    const wrapper = shallow(<AddBook {...props} {...state} /> )
    wrapper.setState({
      title: 'title',
      author: 'author'
    });
    props.updateBook.mockReset();
    wrapper.instance().handleSubmit(event)
    expect(wrapper.instance().state.title).toBe('title');
    expect(wrapper.instance().state.author).toBe('author');
    // expect(props.saveBook).toHaveBeenCalled()
    
  });

  test('should handle image upload', () => {
    const { wrapper, props, state } = setup();
    const event = {
      preventDefault: jest.fn(),
      target: {
        files: ['well firts image link',]
      }
    }
    wrapper.setState({
      title: 'title',
      author: 'author',
      publishedYear: '2012',
      isbn: '9874763528198',
      quantity: '9',
      description: 'description',
      aboutAuthor: 'aboutAothor',
      image: '',
      uploadedImage: event.target.files[0],
    });
    props.saveBook.mockReset();
    wrapper.instance().handleFileChange(event)
    expect(wrapper.instance().state.uploadedImage).toBe('well firts image link');
  });

  test('should return an object with props required by component from store', () => {
    const store = {
      errors: {},
      user: {}
    };
    expect(mapStateToProps({ adminReducer: {} })).toHaveProperty('errors');
  });

  test('should dispatch needed actions', () => {
    const dispatch = jest.fn();
    expect(mapDispatchToProps(dispatch)).toHaveProperty('checkIsbnExist')
    expect(mapDispatchToProps(dispatch)).toHaveProperty('saveBook')
    const { checkIsbnExist, saveBook } = mapDispatchToProps(dispatch);
    checkIsbnExist();
    saveBook();
    expect(dispatch).toHaveBeenCalled();
  })
})

