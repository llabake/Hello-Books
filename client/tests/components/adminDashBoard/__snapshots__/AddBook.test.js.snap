// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add book component should render appropriately 1`] = `
<div>
  <div
    id="banner"
  >
    <div
      className="container form-style"
    >
      <div
        className="row"
      >
        <div
          className="row signup-head"
        >
          <div
            className="col s12"
          >
            <h5
              className="center-align"
            >
               Add Book 
            </h5>
          </div>
        </div>
        <div
          className="col s12 "
        >
          <div
            className="row "
          >
            <form
              className="col s12 signup"
              onSubmit={[Function]}
            >
              <TextInput
                icon="title"
                id="title"
                name="title"
                onChange={[Function]}
                placeholder="Title"
                type="text"
                value=""
              />
              <TextInput
                icon="account_circle"
                id="author"
                name="author"
                onChange={[Function]}
                placeholder="Author"
                type="text"
                value=""
              />
              <TextInput
                icon="date_range"
                id="publishedYear"
                name="publishedYear"
                onChange={[Function]}
                placeholder="Published Year"
                type="text"
                value=""
              />
              <TextInput
                icon="lock"
                id="isbn"
                name="isbn"
                onBlur={[Function]}
                onChange={[Function]}
                placeholder="ISBN"
                type="text"
                value=""
              />
              <TextInput
                icon="date_range"
                id="quantity"
                name="quantity"
                onChange={[Function]}
                placeholder="Quantity"
                type="number"
                value=""
              />
              <TextAreaInput
                icon="short_text"
                id="description"
                name="description"
                onChange={[Function]}
                placeholder="Description"
                value=""
              />
              <TextAreaInput
                icon="short_text"
                id="aboutAuthor"
                name="aboutAuthor"
                onChange={[Function]}
                placeholder="About the Author"
                value=""
              />
              <div
                className="file-field input-field"
                id="image-field"
                style={
                  Object {
                    "marginTop": "35em",
                  }
                }
              >
                <div
                  className="btn"
                  style={
                    Object {
                      "fontSize": "13px",
                      "marginTop": "34px",
                      "width": "40%",
                    }
                  }
                >
                  <span>
                    Upload Image
                  </span>
                  <input
                    accept="image/*"
                    onChange={[Function]}
                    style={
                      Object {
                        "marginTop": "74px",
                      }
                    }
                    type="file"
                  />
                </div>
                <div
                  className="file-path-wrapper"
                >
                  <input
                    className="file-path validate"
                    style={
                      Object {
                        "marginTop": "29px",
                      }
                    }
                    type="text"
                  />
                </div>
              </div>
              <button
                className="waves-effect waves-light btn"
                disabled={true}
                style={
                  Object {
                    "marginTop": "2em",
                  }
                }
                type="submit"
              >
                Add Book
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
